services/gossip/adapter/tcp/direct_transport_test.go

```diff
@@ -36,58 +36,196 @@ func TestDirectTransport_HandlesStartupWithEmptyPeerList(t *testing.T) {
 
-		node1.AddPeer(ctx, address2, config.NewHardCodedGossipPeer(node2.GetServerPort(), "127.0.0.1", hex.EncodeToString(address1)))
-		node2.AddPeer(ctx, address1, config.NewHardCodedGossipPeer(node1.GetServerPort(), "127.0.0.1", hex.EncodeToString(address2)))
+		node1.addPeer(ctx, node2)
+		node2.addPeer(ctx, node1)
 
-		require.True(t, test.Eventually(HARNESS_OUTGOING_CONNECTIONS_INIT_TIMEOUT, func() bool {
-			return len(node1.clientConnections.peers) > 0 && len(node2.clientConnections.peers) > 0
-		}), "expected all outgoing queues to become enabled after successfully connecting to added peers")
+		waitForAllNodesToSatisfy(t,
+			"expected all nodes to have peers added",
+			func(node *nodeHarness) bool { return len(node.transport.clientConnections.peers) > 0 },
+			node1, node2)
 
-		header := (&gossipmessages.HeaderBuilder{
-			Topic:         gossipmessages.HEADER_TOPIC_LEAN_HELIX,
-			RecipientMode: gossipmessages.RECIPIENT_LIST_MODE_BROADCAST,
-		}).Build()
+		waitForAllNodesToSatisfy(t,
+			"expected all outgoing queues to become enabled after successfully connecting to added peers",
+			func(node *nodeHarness) bool { return node.transport.allOutgoingQueuesEnabled() },
+			node1, node2)
 
-		message := &gossipmessages.LeanHelixMessage{
-			Content: []byte{},
-		}
+		node1.requireSendsSuccessfullyTo(t, ctx, node2)
+		node2.requireSendsSuccessfullyTo(t, ctx, node1)
+
+	})
+}
 
-		payloads := [][]byte{header.Raw(), message.Content}
+func TestDirectTransport_SupportsTopologyChangeInRuntime(t *testing.T) {
+	logger := log.DefaultTestingLogger(t)
+	test.WithContext(func(ctx context.Context) {
+		node1 := aNode(ctx, logger)
+		node2 := aNode(ctx, logger)
+		node3 := aNode(ctx, logger)
+		node4 := aNode(ctx, logger)
+
+		waitForAllNodesToSatisfy(t, "server did not start", func(node *nodeHarness) bool { return node.transport.IsServerListening() }, node1, node2, node3, node4)
+
+		firstTopology := aTopologyContaining(node1, node2, node3)
+		node1.transport.UpdateTopology(ctx, firstTopology)
+		node2.transport.UpdateTopology(ctx, firstTopology)
+		node3.transport.UpdateTopology(ctx, firstTopology)
+
+		waitForAllNodesToSatisfy(t,
+			"expected all nodes to have peers added",
+			func(node *nodeHarness) bool { return len(node.transport.clientConnections.peers) > 0 },
+			node1, node2, node3)
+
+		waitForAllNodesToSatisfy(t,
+			"expected all outgoing queues to become enabled after topology change",
+			func(node *nodeHarness) bool { return node.transport.allOutgoingQueuesEnabled() },
+			node1, node2, node3)
+
+		node1.requireSendsSuccessfullyTo(t, ctx, node2)
+		node2.requireSendsSuccessfullyTo(t, ctx, node1)
+		node2.requireSendsSuccessfullyTo(t, ctx, node3)
+
+		secondTopology := aTopologyContaining(node1, node2, node4)
+		node1.transport.UpdateTopology(ctx, secondTopology)
+		node2.transport.UpdateTopology(ctx, secondTopology)
+		node4.transport.UpdateTopology(ctx, secondTopology)
+
+		waitForAllNodesToSatisfy(t,
```

Comment:
1. Since connections are a heavy resource, I think we should have tests to make sure we're not leaking.
2. And, are we justified in allowing connections from clients which are not in our peer list?